pkg/ingester/wal.go

```diff
@@ -0,0 +1,209 @@
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/tsdb/wal"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+var (
+	// shared pool for WALRecords and []logproto.Entries
+	recordPool = newRecordPool()
+)
+
+const walSegmentSize = wal.DefaultSegmentSize * 4
+
+type WALConfig struct {
+	Enabled            bool          `yaml:"enabled"`
+	Dir                string        `yaml:"dir"`
+	Recover            bool          `yaml:"recover"`
+	CheckpointDuration time.Duration `yaml:"checkpoint_duration"`
+	FlushOnShutdown    bool          `yaml:"flush_on_shutdown"`
+}
+
+func (cfg *WALConfig) Validate() error {
+	if cfg.Enabled && cfg.CheckpointDuration < 1 {
+		return errors.Errorf("invalid checkpoint duration: %v", cfg.CheckpointDuration)
+	}
+	return nil
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *WALConfig) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.Dir, "ingester.wal-dir", "wal", "Directory to store the WAL and/or recover from WAL.")
+	f.BoolVar(&cfg.Enabled, "ingester.wal-enabled", false, "Enable writing of ingested data into WAL.")
+	f.BoolVar(&cfg.Recover, "ingester.recover-from-wal", false, "Recover data from existing WAL irrespective of WAL enabled/disabled.")
+	f.DurationVar(&cfg.CheckpointDuration, "ingester.checkpoint-duration", 5*time.Minute, "Interval at which checkpoints should be created.")
+	f.BoolVar(&cfg.FlushOnShutdown, "ingester.flush-on-shutdown", false, "When WAL is enabled, should chunks be flushed to long-term storage on shutdown.")
+}
+
+// WAL interface allows us to have a no-op WAL when the WAL is disabled.
+type WAL interface {
+	// Log marshalls the records and writes it into the WAL.
+	Log(*WALRecord) error
+	// Stop stops all the WAL operations.
+	Stop()
+}
+
+type noopWAL struct{}
+
+func (noopWAL) Log(*WALRecord) error { return nil }
+func (noopWAL) Stop()                {}
+
+type walWrapper struct {
+	cfg        WALConfig
+	wal        *wal.WAL
+	metrics    *ingesterMetrics
+	seriesIter SeriesIter
+
+	wait sync.WaitGroup
+	quit chan struct{}
+}
+
+// newWAL creates a WAL object. If the WAL is disabled, then the returned WAL is a no-op WAL.
+func newWAL(cfg WALConfig, registerer prometheus.Registerer, metrics *ingesterMetrics, seriesIter SeriesIter) (WAL, error) {
+	if !cfg.Enabled {
+		return noopWAL{}, nil
+	}
+
+	tsdbWAL, err := wal.NewSize(util.Logger, registerer, cfg.Dir, walSegmentSize, false)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &walWrapper{
+		cfg:        cfg,
+		quit:       make(chan struct{}),
+		wal:        tsdbWAL,
+		metrics:    metrics,
+		seriesIter: seriesIter,
+	}
+
+	w.wait.Add(1)
+	go w.run()
+	return w, nil
+}
+
+func (w *walWrapper) Log(record *WALRecord) error {
+	if record == nil || (len(record.Series) == 0 && len(record.RefEntries) == 0) {
+		return nil
+	}
+	select {
+	case <-w.quit:
+		return nil
```

Comment:
Q: should we return some error if canceled via `w.quit` rather than returning just `nil`? to differentiate from normal success case?

Reply:
I don't think quitting is an error -- I made the conscious decision to not wait for a checkpoint before shutdown so we can fail fast. In this case we'll still have the previous checkpoint + WAL to recover from.